compiler/packages/babel-plugin-react-compiler/src/Transform/TransformFire.ts

```diff
@@ -0,0 +1,625 @@
+    const deleteInstrs = new Set<InstructionId>();
+    for (const instr of block.instructions) {
+      const {value, lvalue} = instr;
+      if (
+        value.kind === 'CallExpression' &&
+        isUseEffectHookType(value.callee.identifier) &&
+        value.args.length > 0 &&
+        value.args[0].kind === 'Identifier'
+      ) {
+        const lambda = context.getFunctionExpression(
+          value.args[0].identifier.id,
+        );
+        if (lambda != null) {
+          const capturedCallees =
+            visitFunctionExpressionAndPropagateFireDependencies(
+              lambda,
+              context,
+              true,
+            );
+
+          // Add useFire calls for all fire calls in found in the lambda
+          const newInstrs = [];
+          for (const [
+            fireCalleePlace,
+            fireCalleeInfo,
+          ] of capturedCallees.entries()) {
+            if (!context.hasCalleeWithInsertedFire(fireCalleePlace)) {
+              context.addCalleeWithInsertedFire(fireCalleePlace);
+              const {loadUseFireInstr, useFirePlace} =
+                makeLoadUseFireInstruction(fn.env);
+              const {loadFireCalleeInstr, loadedFireCallee} =
+                makeLoadFireCalleeInstruction(
+                  fn.env,
+                  fireCalleeInfo.capturedCalleeIdentifier,
+                );
+              const {callUseFireInstr, useFireCallResultPlace} =
+                makeCallUseFireInstruction(
+                  fn.env,
+                  useFirePlace,
+                  loadedFireCallee,
+                );
+              const storeUseFireInstr = makeStoreUseFireInstruction(
+                fn.env,
+                useFireCallResultPlace,
+                fireCalleeInfo.fireFunctionBinding,
+              );
+              newInstrs.push(
+                loadUseFireInstr,
+                loadFireCalleeInstr,
+                callUseFireInstr,
+                storeUseFireInstr,
+              );
+
+              // We insert all of these instructions before the useEffect is loaded
+              const loadUseEffectInstrId = context.getLoadGlobalInstrId(
+                value.callee.identifier.id,
+              );
+              if (loadUseEffectInstrId == null) {
+                context.pushError({
+                  loc: value.loc,
+                  description: null,
+                  severity: ErrorSeverity.Invariant,
+                  reason: '[InsertFire] No LoadGlobal found for useEffect call',
+                  suggestions: null,
+                });
+                continue;
+              }
+              rewriteInstrs.set(loadUseEffectInstrId, newInstrs);
+            }
+          }
+        }
+      } else if (
+        value.kind === 'CallExpression' &&
+        value.callee.identifier.type.kind === 'Function' &&
+        value.callee.identifier.type.shapeId === BuiltInFireId &&
+        context.inUseEffectLambda()
```

Comment:
could maybe move this check down into the block, and throw an InvalidReact error if we're not in an effect lambda

Reply:
I handle this error in a future stacked diff. The reason I don't emit errors in this pass is because we only traverse functions that are passed to effects. I remove all of the fire calls we can reach by doing that and then do a post-pass where we error on any lingering ones.

We can't eagerly remove them because we don't know if we should traverse a function expression until we see it is passed to a useEffect lambda